python/src/main.cpp

```diff
@@ -0,0 +1,906 @@
+             py::arg("elemsize") = 4, py::arg("elempack") = 1, py::arg("allocator") = nullptr)
+
+        .def(py::init<const Mat&>(), py::arg("m"))
+
+        .def(py::init([](py::buffer const b) {
+                 py::buffer_info info = b.request();
+                 if (info.ndim > 3)
+                 {
+                     throw std::runtime_error("Incompatible buffer dims");
+                 }
+
+                 size_t elemsize = 4u;
+                 if (info.format == py::format_descriptor<double>::format())
+                 {
+                     elemsize = 8u;
+                 }
+                 if (info.format == py::format_descriptor<float>::format() || info.format == py::format_descriptor<int>::format())
+                 {
+                     elemsize = 4u;
+                 }
+                 else if (info.format == "e")
+                 {
+                     elemsize = 2u;
+                 }
+                 else if (info.format == py::format_descriptor<int8_t>::format() || info.format == py::format_descriptor<uint8_t>::format())
+                 {
+                     elemsize = 1u;
+                 }
+
+                 Mat* v = nullptr;
+                 if (info.ndim == 1)
+                 {
+                     v = new Mat((int)info.shape[0], info.ptr, elemsize);
+                 }
+                 else if (info.ndim == 2)
+                 {
+                     v = new Mat((int)info.shape[1], (int)info.shape[0], info.ptr, elemsize);
+                 }
+                 else if (info.ndim == 3)
+                 {
+                     v = new Mat((int)info.shape[2], (int)info.shape[1], (int)info.shape[0], info.ptr, elemsize);
+                 }
+                 return v;
+             }),
+             py::arg("array"))
+        .def_buffer([](Mat& m) -> py::buffer_info {
+            std::string format = get_mat_format(m);
+            std::vector<ssize_t> shape;
+            std::vector<ssize_t> strides;
+            if (m.dims == 1)
+            {
+                shape.push_back(m.w);
+                strides.push_back(m.elemsize);
+            }
+            else if (m.dims == 2)
+            {
+                shape.push_back(m.h);
+                shape.push_back(m.w);
+                strides.push_back(m.w * m.elemsize);
+                strides.push_back(m.elemsize);
+            }
+            else if (m.dims == 3)
+            {
+                shape.push_back(m.c);
+                shape.push_back(m.h);
+                shape.push_back(m.w);
+                strides.push_back(m.cstep * m.elemsize);
+                strides.push_back(m.w * m.elemsize);
+                strides.push_back(m.elemsize);
+            }
+            return py::buffer_info(
+                m.data,     /* Pointer to buffer */
+                m.elemsize, /* Size of one scalar */
+                format,     /* Python struct-style format descriptor */
```

Comment:
elempack info lost
It seems numpy does not have the elempack semantic
what about another dimension for elempack != 1 ? [c, h, w, elempack]